arangod/Aql/WindowExecutor.cpp

```diff
@@ -0,0 +1,439 @@
+};
+
+AccuWindowExecutor::~AccuWindowExecutor() {}
+
+void AccuWindowExecutor::initializeCursor() {
+  resetAggregators();
+}
+
+std::tuple<ExecutorState, NoStats, AqlCall> AccuWindowExecutor::produceRows(
+    AqlItemBlockInputRange& inputRange, OutputAqlItemRow& output) {
+  // This block is passhthrough.
+  static_assert(Properties::allowsBlockPassthrough == BlockPassthrough::Enable,
+                "For WINDOW with passthrough to work, there must be "
+                "exactly enough space for all input in the output.");
+
+  // simple optimization for culmulative SUM and the likes
+  while (inputRange.hasDataRow()) {
+    // So there will always be enough place for all inputRows within
+    // the output.
+    TRI_ASSERT(!output.isFull());
+    auto [state, input] = inputRange.nextDataRow(AqlItemBlockInputRange::HasDataRow{});
+    TRI_ASSERT(input.isInitialized());
+    applyAggregators(input);
+    produceOutputRow(input, output, /*reset*/ false);
+  }
+
+  // Just fetch everything from above, allow overfetching
+  AqlCall upstreamCall{};
+  return {inputRange.upstreamState(), NoStats{}, upstreamCall};
+}
+
+/**
+ * @brief Skip Rows
+ *   We need to consume all rows from the inputRange
+ *
+ * @param inputRange  Data from input
+ * @param call Call from client
+ * @return std::tuple<ExecutorState, NoStats, size_t, AqlCall>
+ */
+auto AccuWindowExecutor::skipRowsRange(AqlItemBlockInputRange& inputRange, AqlCall& call)
+    -> std::tuple<ExecutorState, NoStats, size_t, AqlCall> {
+  // we do not keep any state
+  AqlCall upstreamCall{};
+  return {inputRange.upstreamState(), NoStats{}, call.getSkipCount(), upstreamCall};
+}
+
+[[nodiscard]] auto AccuWindowExecutor::expectedNumberOfRowsNew(
+    AqlItemBlockInputRange const& input, AqlCall const& call) const noexcept -> size_t {
+  if (input.finalState() == ExecutorState::DONE) {
+    // For every input row we produce a new row.
+    auto estOnInput = input.countDataRows();
+    return std::min(call.getLimit(), estOnInput);
+  }
+  // We do not know how many more rows will be returned from upstream.
+  // So we can only overestimate
+  return call.getLimit();
+}
+
+// -------------- WindowExecutor --------------
+
+WindowExecutor::WindowExecutor(Fetcher& fetcher, Infos& infos)
+    : BaseWindowExecutor(infos){};
+
+WindowExecutor::~WindowExecutor() {}
+
+ExecutorState WindowExecutor::consumeInputRange(AqlItemBlockInputRange& inputRange) {
+  const RegisterId rangeRegister = _infos.rangeRegister();
+  QueryWarnings& qc = _infos.warnings();
+  WindowBounds const& b = _infos.bounds();
+
+  while (inputRange.hasDataRow()) {
+    auto [state, input] = inputRange.nextDataRow(AqlItemBlockInputRange::HasDataRow{});
+    TRI_ASSERT(input.isInitialized());
+    _rows.push_back(input);
+
+    if (rangeRegister != RegisterPlan::MaxRegisterId) {
+      AqlValue val = input.getValue(rangeRegister);
+      _windowRows.emplace_back(b.calcRow(val, qc));
+    }
```

Comment:
We can avoid copying the SharedAqlItemBlockPtr here:
```suggestion
    if (rangeRegister != RegisterPlan::MaxRegisterId) {
      AqlValue val = input.getValue(rangeRegister);
      _windowRows.emplace_back(b.calcRow(val, qc));
    }
    
    _rows.push_back(std::move(input));
```